### ISSUE TITLE
Non-standard behavior in MPI_File_open

### ISSUE THREAD

**Issue by nasailja** [open-mpi/ompi#862]
This program does not destroy the file on open, even though it should:
```
#include "stdio.h"
#include "mpi.h"
int main(int argc, char* argv[]){
	char buf[2] = {'b', '\n'};
	#ifdef USE_STDIO
	FILE* f1 = fopen("a", "w");
	if (f1 == NULL) { return 1; };
	if (fseek(f1, 0, SEEK_END) != 0) { return 2; };
	if (fwrite(buf, 1, 2, f1) != 2) { return 3; };
	#else
	if (MPI_Init(&argc, &argv) != MPI_SUCCESS) { return 1; }
	MPI_File f2;
	if (MPI_File_open(
		MPI_COMM_WORLD,
		"a",
		MPI_MODE_WRONLY,
		MPI_INFO_NULL,
		&f2
	) != MPI_SUCCESS) { return 2; }
	if (MPI_File_seek(
		f2,
		0,
		MPI_SEEK_END
	) != MPI_SUCCESS) { return 3; }
	if (MPI_File_write(
		f2,
		buf,
		2,
		MPI_CHAR,
		MPI_STATUS_IGNORE
	) != MPI_SUCCESS) { return 4; }
	MPI_Finalize();
	#endif
	return 0;
}

mpiexec --version
mpiexec (OpenRTE) 1.8.8

mpicc filewritebug.c -o filewritebug && echo a > a && mpiexec -n 1 ./filewritebug && echo $? && cat a
0
a
b

mpicc -DUSE_STDIO filewritebug.c -o filewritebug && echo a > a && mpiexec -n 1 ./filewritebug && echo $? && cat a
0
b
```
As far as I can tell (http://en.cppreference.com/w/cpp/io/c/fopen mode "w", action if file already exists) MPI_MODE_WRONLY must destroy the file upon open which is what happens when replacing MPI with stdio in above program.

**Comment by ggouaillardet** [comment]
@nasailja this is what the standard says (chapter 13.2, page 495 of the mpi 3.1 standard)

```
The modes MPI_MODE_RDONLY, MPI_MODE_RDWR, MPI_MODE_WRONLY,
MPI_MODE_CREATE, and MPI_MODE_EXCL have identical semantics to their POSIX counterparts
```

my understanding is that POSIX means ```open``` and __not__ ```fopen```

so if you rewrite the test program with posix calls, both posix and mpi versions produce the same results

```c
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "stdio.h"
#include "mpi.h"
int main(int argc, char* argv[]){
    char buf[2] = {'b', '\n'};
    #ifdef USE_POSIX
    int f1 = open("a", O_WRONLY);
    if (f1 < 0) { return 1; };
    if (lseek(f1, 0, SEEK_END) < 0) { return 2; };
    if (write(f1, buf, 2) != 2) { return 3; };
    #else
    if (MPI_Init(&argc, &argv) != MPI_SUCCESS) { return 1; }
    MPI_File f2;
    if (MPI_File_open(
        MPI_COMM_WORLD,
        "a",
        MPI_MODE_WRONLY,
        MPI_INFO_NULL,
        &f2
    ) != MPI_SUCCESS) { return 2; }
    if (MPI_File_seek(
        f2,
        0,
        MPI_SEEK_END
    ) != MPI_SUCCESS) { return 3; }
    if (MPI_File_write(
        f2,
        buf,
        2,
        MPI_CHAR,
        MPI_STATUS_IGNORE
    ) != MPI_SUCCESS) { return 4; }
    MPI_Finalize();
    #endif
    return 0;
}
```

i just realized there is no ```MPI_MODE_TRUNC``` equivalent to posix ```O_TRUNC```, but here is not the right place to discuss the standard

**Comment by nasailja** [comment]
True but shouldn't MPI_File_open be called MPI_Open if it were supposed to be equated with open instead of fopen? Fopen also exists in POSIX according to this http://pubs.opengroup.org/onlinepubs/009695399/functions/fopen.html

**Comment by ggouaillardet** [comment]
@jsquyres @bosilca could you please comment ?
should MPI_File_open behave like open or fopen ?
@nasailja MPI IO subroutines start with MPi_File_, so you cannot assume fopen should be used instead of open only because of the name

**Comment by nasailja** [comment]
Don't they start with MPI_File_ because anything MPI can do IO wise operates on files and not FIFOs, sockets, etc.? If yes then you might see why it's easy to (make the mistake of?) check fopen instead of open when trying to figure out how MPI_File_open behaves because fopen also only operates on files and not FIFOs, sockets, etc.

**Comment by ggouaillardet** [comment]
@nasailja i will let George or Jeff comment on that.
unix open does not work with sockets nor non named pipes.

**Comment by nasailja** [comment]
Ah it has to be a named pipe.

**Comment by bosilca** [comment]
Based on my understanding of the MPI IO chapter the only way to truncate a file is by explicitly using MPI_FILE_SET_SIZE. Therefore I would not expect MPI_MODE_WRONLY to destroy and then recreate the file. In addition, open is the only function that takes the action hint as an int. All in all I would expect the MPI flags to be equivalent to the open POSIX flags.

Regarding the naming scheme, in MPI things are specialized, which explains why MPI_FILE* functions only work on files.

@edgargabriel can you comment?

PS: What would it means in a parallel environment to open a named pipe ? If you open independent pipes you lose the collective context of the call.

**Comment by nasailja** [comment]
`What would it means in a parallel environment to open a named pipe ?`
Maybe each rank could write to stdin of the same rank of another program started with mpiexec? E.g. `mpiexec -n N ./a | mpiexec -n N ./b | ...`

**Comment by edgargabriel** [comment]
you are right, just setting MPI_MODE_WRONLY should not erase an existing file, unless MPI_MODE_CREATE has been set in addition. The latter one will however delete an existing file.

**Comment by nasailja** [comment]
According to opengroup.org O_CREAT has no effect if the file already exists, unless O_EXCL is also given in which case open fails.

**Comment by edgargabriel** [comment]
yes, after re-reading you are right. WE anyway just translate the flags to their posix versions, so it should be correct.
